chore: restructure WidgetsEditor

marks0351 opened the pull request:
## Description
In this PR we are restructuring the WidgetsEditor into multiple components so that it is easy to assess changes.
currently the widgets editor just does too much at one place to even understand what is happening.

This is also the first step in trying to decrease the number of places where layout system type is being used to manipulate a functionality.

Fixes #`Issue Number`  
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results
<!-- This is an auto-generated comment: Cypress test results  -->
> [!IMPORTANT]  
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/8321642656>
> Commit: `afbe7fc015e4d7b098d5fb68db8eb17a730fde71`
> Cypress dashboard url: <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=8321642656&attempt=2" target="_blank">Click here!</a>
> All cypress tests have passed 🎉🎉🎉

<!-- end of auto-generated comment: Cypress test results  -->















<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
    - Introduced a comprehensive layout for the widget editor, including header, content, and footer areas.
    - Added `WidgetEditorContent` for orchestrating the main content area with navigation and layout-based canvas editor.
    - New `LayoutSystemBasedPageViewer` and `NavigationAdjustedPageViewer` components for enhanced page display and navigation within the layout system.
    - Implemented `WidgetEditorContentWrapper` for managing content layout and interactions within the widget editor.
    - Added `WidgetEditorNavigation` for improved navigation preview.

- **Refactor**
    - Updated import paths across various components to reflect new project structure.
    - Significant restructuring of the `WidgetsEditor` component for better efficiency and dependency management.

- **Documentation**
    - Added descriptive comments for the `CodeModeTooltip` component.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

riodeuno commented on app/client/src/pages/Editor/WidgetsEditor/WidgetEditorSkeleton.tsx:
```diff
@@ -0,0 +1,32 @@
+import EditorContextProvider from "components/editorComponents/EditorContextProvider";
```
@marks0351 Could you add a comment to describe what this component refers to in the product?

riodeuno commented on app/client/src/pages/Editor/WidgetsEditor/WidgetEditorSkeleton.tsx:
```diff
@@ -0,0 +1,32 @@
+  const isAppSettingsPaneWithNavigationTabOpen =
+    appState === EditorState.SETTINGS && isNavigationSelectedInSettings;
+  return (
+    <EditorContextProvider renderMode="CANVAS">
```
`RenderModes.CANVAS`? Instead of a string?

KelvinOm replied:
There is a minor sentiment about this. I think we are overusing enums. TS is smart enough to understand which type needs to be use. Also, all these files get into the final bundle. I think it's worth considering how we can move towards using union types and regular strings.

riodeuno reviewed (commented):
@marks0351 
The changes look fine to me. Thanks for picking this up and cleaning up the codebase.
A few comments
1. All of RabbitAI's comments make sense to me.
2. Can we use absolute import paths if we're importing from an external module. This is really helpful in refactors.
3. Now that we have separated the components. Can we add component/file level comments explaining why the component exists?

marks0351 commented:
> @marks0351 The changes look fine to me. Thanks for picking this up and cleaning up the codebase. A few comments
> 
> 1. All of RabbitAI's comments make sense to me.
> 2. Can we use absolute import paths if we're importing from an external module. This is really helpful in refactors.
> 3. Now that we have separated the components. Can we add component/file level comments explaining why the component exists?

@riodeuno I have addressed some comments of the ai, some are not feasible like grouping multiple imports from different selectors, I really dont want to rework the current structure and bloat the scope of this initiative. I have added comments as well giving a brief understanding of what the core features of the respective components are.

marks0351 commented:
/build-deploy-preview skip-tests=true

KelvinOm commented on app/client/src/pages/Editor/WidgetsEditor/components/LayoutSystemBasedPageViewer.tsx:
```diff
@@ -0,0 +1,84 @@
+import SnapShotBannerCTA from "pages/Editor/CanvasLayoutConversion/SnapShotBannerCTA";
+import React from "react";
+import styled from "styled-components";
+import MainContainerWrapper from "./MainContainerWrapper";
+import OverlayCanvasContainer from "layoutSystems/common/WidgetNamesCanvas";
+import { AppSettingsTabs } from "pages/Editor/AppSettingsPane/AppSettings";
+import { useSelector } from "react-redux";
+import {
+  getCanvasWidth,
+  getCurrentPageId,
+  previewModeSelector,
+} from "selectors/editorSelectors";
+import { protectedModeSelector } from "selectors/gitSyncSelectors";
+import { getSnapshotUpdatedTime } from "selectors/autoLayoutSelectors";
+import { getReadableSnapShotDetails } from "layoutSystems/autolayout/utils/AutoLayoutUtils";
+import {
+  getAppSettingsPaneContext,
+  getIsAppSettingsPaneWithNavigationTabOpen,
+} from "selectors/appSettingsPaneSelectors";
+import { useCurrentAppState } from "pages/Editor/IDE/hooks";
+import { EditorState } from "@appsmith/entities/IDE/constants";
+import {
+  LayoutSystemFeatures,
+  useLayoutSystemFeatures,
+} from "layoutSystems/common/useLayoutSystemFeatures";
```
Eslint can do this automatically for us, we just need to set up the rule. I would skip it for now.

KelvinOm commented on app/client/src/pages/Editor/WidgetsEditor/WidgetEditorContainer.tsx:
```diff
@@ -0,0 +1,37 @@
+  return (
+    <EditorContextProvider renderMode={RenderModes.CANVAS}>
+      <div className="relative flex flex-row h-full w-full overflow-hidden">
+        <div
```
We can get rid of one of these divs

KelvinOm commented on app/client/src/pages/Editor/WidgetsEditor/components/PropertyPaneWrapper.tsx:
```diff
@@ -3,10 +3,12 @@ import PropertyPaneSidebar from "components/editorComponents/PropertyPaneSidebar
+import { CreateNewModal } from "pages/Editor/IDE/RightPane/components/CreateNewModal";
 
 /**
  * OldName: PropertyPaneContainer
```
Seems this comment doesn't seem to make sense now.

KelvinOm commented on app/client/src/pages/Editor/WidgetsEditor/components/WidgetEditorNavigation.tsx:
```diff
@@ -0,0 +1,71 @@
+export const useNavigationPreviewHeight = () => {
+  const [navigationHeight, setNavigationHeight] = useState(0);
+  const navigationPreviewRef = useRef(null);
+  const isPreviewMode = useSelector(combinedPreviewModeSelector);
```
What if we use `useSelector` only in the `WidgetEditorContent` and just pass the props into other components. It will work faster and code readability will get better.

marks0351 commented:
@KelvinOm thanks for the comments, I will be addressing cleanup of dom nodes, state management as part of a future task once I find some QA bandwith since these components could be responsible for proper functioning of endless number of use cases which would require manual regression. for now the goal is to modularise so that each component can be cleaned up further individually with minimal radius of impact.
